### PR TITLE
[FSDP] Enhance test for checkpoint

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75236
* #75227

Make sure that the param names are such as we can load from checkpoint
into a local model.

Differential Revision: [D35377774](https://our.internmc.facebook.com/intern/diff/D35377774/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 1]
```diff
@@ -324,12 +326,30 @@ def _create_module():
...
+        local, _ = _create_module(wrap_fsdp=False)
+        for param in local.parameters():
+            with torch.no_grad():
+                param.zero_()
+
+        with fsdp.summon_full_params():
+            for (p1, p2) in zip(fsdp.parameters(), local.parameters()):
+                self.assertNotEqual(p1, p2)
+
+        local.load_state_dict(deepcopy(state_dict))
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_state_dict.py:348]: Curious, what is the need for the `deepcopy()`?

**Review by awgu — APPROVED** [review]
Looks good!

File: test/distributed/fsdp/test_fsdp_state_dict.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_state_dict.py:348]: It's used in a few other tests in this file, and I think the aim is to better simulate a scenario where we don't have any access to the original state_dict and instead recover it from a file. Though I think doing an actual torch.save / torch.load using tempfiles would a better test here.

**Pull request closed without merging**